internal/transaction/paxos.go

```diff
@@ -0,0 +1,72 @@
+// MIT License. Copyright (c) 2020 CQFN
+// https://github.com/cqfn/degitx/blob/master/LICENSE
+
+package transaction
+
+import (
+	"context"
+)
+
+// Ballot number of proposals in Paxos algorithm
+type Ballot uint16
+
+// Proposal for Paxos protocol
+type Proposal struct {
+	// Ballot number of the proposal
+	Ballot
+	// Proposer ID, it should be unique number for each
+	// proposer participating in the transaction.
+	Proposer uint16
+}
+
+// Compare two proposals
+func (p Proposal) Compare(o Proposal) int {
+	var r int
+	r = int(p.Ballot - o.Ballot)
+	if r == 0 {
+		r = int(p.Proposer - o.Proposer)
+	}
+	return r
+}
+
+// PxAcceptor - Paxos acceptor API for proposer
+type PxAcceptor interface {
+	// Prepare takes Paxos 1A message. The proposer chooses ballot number
+	// that it believes to be larger than any ballot number for which
+	// phase 1 has been performed, and sends it to every acceptor for prepare.
+	Prepare(ctx context.Context, p Proposal) error
+
+	// Accept takes Paxos 2A message. The proposer uses ballot number it has
+	// already prepared and acceptor promised to reject all proposals fewer than
+	// prepared one. It tries to accept a vote by proposal. If the bigger ballot
+	// number was prepared by acceptor between these two operations, acceptor may
+	// respond with reject.
+	Accept(ctx context.Context, p Proposal, v Vote) error
+}
+
+// PxProposer - Paxos proposer API for acceptor
+type PxProposer interface {
+	// Promise is a 1B message from acceptor.
+	// was prepared succesffully, it includes a proposal ballot
+	// number. Promise means that acceptor promises to proposer
+	// to reject all next messages if ballot number of such message
+	// are less than ballot number of promise.
+	Promise(ctx context.Context, b Ballot) (max Ballot, acc Ballot, v Vote, err error)
+
```

Comment:
Now `Acceptor` has no way to notify `Proposer` what `value` was already chosen. I think we steel need the second method. If we want to stay with a single method, we have to pass nullable `Value`. Respond to prepare should contains a chosen value. [see](https://youtu.be/JEpsBg0AO6o?t=1049).